### PR TITLE
[Feature-13419][Remote Logging] Add support for writing task logs to S3

### LINKED ISSUE #13419: [Feature][Remote Logging] Add support for writing task logs to S3
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

* This is a sub-task of #13017 
* Add support for writing task logs to S3

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by rickchengx**
## Purpose of the pull request
* close: #13419 
* A sub-task of #13017 

## Brief change log
Add support for writing task logs to S3

## Verify this pull request
UT

### CODE REVIEW

**Comment by Radeity** [comment]
Hi, @rickchengx , may i ask why we have to create new instance of `RemoteLogHandler` to send remote log every time, can we just create a single instance to handle that?

**Comment by rickchengx** [comment]
> Hi, @rickchengx , may i ask why we have to create new instance of `RemoteLogHandler` to send remote log every time, can we just create a single instance to handle that?

Hi, @Radeity Thanks a lot for your comment. I've changed to singleton pattern.

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/log/remote/S3RemoteLogHandler.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            int readLen = 0;
+            while ((readLen = s3is.read(readBuf)) > 0) {
+                fos.write(readBuf, 0, readLen);
+            }
+            s3is.close();
+            fos.close();
+        } catch (Exception e) {
+            log.error("error while getting remote log on S3 {} to {}", objectName, logPath, e);
+        }
+    }
```
Comment (caishunfeng) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/log/remote/S3RemoteLogHandler.java]: It's better to use `try-with-resource` way.

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/log/remote/S3RemoteLogHandler.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        accessKeySecret = readAccessKeySecret();
+        region = readRegion();
+        bucketName = readBucketName();
+        endPoint = readEndPoint();
+        s3Client = buildS3Client();
+        checkBucketNameExists(bucketName);
+    }
+
+    protected AmazonS3 buildS3Client() {
+        if (!StringUtils.isEmpty(endPoint)) {
```
Comment (caishunfeng) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/log/remote/S3RemoteLogHandler.java]:
```suggestion
        if (StringUtils.isNotEmpty(endPoint)) {
```

**Comment by rickchengx** [comment]
Hi, @caishunfeng , thanks a lot for your comment. I've changed it according to your suggestions and please help review this again.

**Review by EricGao888 — DISMISSED** [review]
Would u like to double check when available? @caishunfeng @Radeity

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/log/remote/RemoteLogHandlerFactory.java [hunk 3]
```diff
@@ -21,19 +21,25 @@
...
+        if ("OSS".equals(target)) {
+            return OssRemoteLogHandler.getInstance();
+        } else if ("S3".equals(target)) {
+            return S3RemoteLogHandler.getInstance();
         }
-        OssRemoteLogHandler ossRemoteLogHandler = new OssRemoteLogHandler();
-        ossRemoteLogHandler.init();
-        return ossRemoteLogHandler;
+
+        log.error("No suitable remote logging target for {}", target);
```
Comment (Radeity) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/log/remote/RemoteLogHandlerFactory.java:42]: Some tiny suggestion, if we write an error log here, do we have to also write an error log which may be close in meaning when judging `remoteLogHandler == null`?
```java
if (remoteLogHandler == null) {
    log.error("remote log handler is null");
    return;
}
```

**Review by Radeity — DISMISSED** [review]
LGTM 👍

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/log/remote/RemoteLogHandlerFactory.java (hunk 3 above)
Comment (rickchengx) [reply dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/log/remote/RemoteLogHandlerFactory.java:42]: Sure

**Review by EricGao888 — APPROVED** [review]

**Pull request merged**